may-2018-reenabled-opcodes.md

```diff
@@ -0,0 +1,428 @@
+|OP_BIN2NUM |129    |0x81|x             |out     |convert byte sequence *x* into a numeric value                    |
+
+Splice operations: `OP_CAT`, `OP_SPLIT`**
+
+Bitwise logic: `OP_AND`, `OP_OR`, `OP_XOR`
+
+Arithmetic: `OP_DIV`, `OP_MOD`
+
+New operations: 
+* `x OP_BIN2NUM -> n`, convert a byte sequence `x` into a numeric value
+* `n m OP_NUM2BIN -> out`, convert a numeric value `n` into a byte sequence of length `m`
+
+Further discussion of the purpose of these new operations can be found below under *bitwise operations*.
+
+** A new operation, `OP_SPLIT`, has been designed as a replacement for `OP_SUBSTR`, `OP_LEFT`and `OP_RIGHT`. 
+The original operations can be implemented with varying combinations of `OP_SPLIT`, `OP_SWAP` and `OP_DROP`.
+
+
+## <a name="data-types"></a>Script data types
+
+It should be noted that in script operation data values on the stack are interpreted as either byte sequences  
+or numeric values.  **All data on the stack is interpreted as a byte sequence unless specifically stated as being interpreted 
+as a numeric value.**
+
+The numeric value type has specific limitations:
+1. The used encoding is little endian with an explicit sign bit (the highest bit of the last byte).
+2. They cannot exceed 4 bytes in length.
+3. They must be encoded using the shortest possible byte length (no zero padding)
+    3. There is one exception to rule 3: if there is more than one byte and the most significant bit of the 
+        second-most-significant-byte is set it would conflict with the sign bit. In this case a single 0x00 or 0x80 byte is allowed
+        to the left.
+4. Zero is encoded as a zero length byte sequence. Single byte positive or negative zero (0x00 or 0x80) are not allowed.
+    
+The new opcode `x OP_BIN2NUM -> out` can be used convert a byte sequence into a numeric value where required.
+
+The new opcode `x n OP_NUM2BIN` can be used to convert a numeric value into a zero padded byte sequence of length `n` 
+whilst preserving the sign bit.
+
+**Endian notation**
+
+For human readability where hex strings are presented in this document big endian notation is used.  That is: 0x0100 represents 
+decimal 256, not decimal 1.
+
+
+## Risks and philosophical approach during design
+
+In general the approach taken is a minimalist one in order limit edge cases as much as possible.  Where it is possible
+for a primitive op code used in conjunction with existing op codes to be combined to produce several more complex 
+operations that is preferred over a set of more complex op codes.  Input conditions that create ambiguous or undefined 
+behaviour should fail fast.  
+
+Each op code should be examined for the following risk conditions and mitigating behaviour defined explicitly:
+* Operand byte length mismatch.  Where it would be normally expected that two operands would be of matching byte lengths
+  the resultant behaviour should be defined.
```

Comment:
This is an example where "should" versus "must" ambiguity is a problem. Let's say an operator requires two operands of equal length, but doesn't specify a default behavior if they're not. What happens if for some reason it gets operands of unequal lengths? Client A defines its result to use the shorter length and truncate the longer one's length. Client B defines its result to pad or otherwise expand the shorter operand to match the longer length. Client C simply fails the operation. And now we have a mess.

I realize that each operator below has detailed test cases for each such scenario. This paragraph is discussing the philosophy behind implementing the opcodes. It should be clear in the policy that such scenarios must, not should, have a defined behavior.

Reply:
section removed